text/0030-protocol-messages.md

```diff
@@ -0,0 +1,243 @@
++ Start date: 2020-04-27
++ RFC PR: [iotaledger/bee-rfcs#30](https://github.com/iotaledger/bee-rfcs/pull/30)
++ Bee issue: [iotaledger/bee#70](https://github.com/iotaledger/bee/issues/70)
+
+# Summary
+
+This RFC introduces the IOTA protocol messages.
+
+# Motivation
+
+To be able to take part in the IOTA networks, Bee nodes need to implement the exact same protocol presented in this RFC
+and currently being used by [IRI](https://github.com/iotaledger/iri) nodes and
+[HORNET](https://github.com/gohornet/hornet) nodes.
+
+# Detailed design
+
+## `Message` trait
+
+The `Message` trait is protocol agnostic and only provides serialization and deserialization to and from bytes buffers.
+It should not be used as is but rather be paired with a higher layer - like a type-length-value encoding - and as such
+does not provide any safety check on inputs/outputs.
+
+```rust
+/// A trait describing the behavior of a message.
+pub(crate) trait Message {
+    /// The unique identifier of the message within the protocol.
+    const ID: u8;
+
+    /// Returns the size range of the message as it can be compressed.
+    fn size_range() -> Range<usize>;
+
+    /// Deserializes a bytes buffer into a message.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be within the range returned by the `size_range` method.
+    fn from_bytes(bytes: &[u8]) -> Self;
+
+    /// Returns the size of the message.
+    fn size(&self) -> usize;
+
+    /// Serializes a message into a bytes buffer.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be equal to the one returned by the `size` method.
+    fn into_bytes(self, bytes: &mut [u8]);
+}
+```
+
+**Notes**:
+- `into_bytes` does not allocate a buffer because the following TLV protocol implies concatenating a header meaning
+  another allocation. Since this is a hot path, a slice of an already allocated buffer for both the header and payload
+  is expected; hence, limiting the amount of allocation to the bare minimum;
+- `from_bytes`/`into_bytes` panic if incorrectly used, only the following safe TLV module should directly use them;
+
+## Type-length-value protocol
+
+The [type-length-value](https://en.wikipedia.org/wiki/Type-length-value) module is a safe layer on top of the messages.
+It allows serialization/deserialization to/from bytes ready to be sent/received to/from a transport layer by prepending
+or reading a header containing the type and length of the payload.
+
+### Header
+
+```rust
+/// A header for the type-length-value encoding.
+pub(crate) struct Header {
+    /// Type of the message.
+    pub(crate) message_type: u8,
+    /// Length of the message.
+    pub(crate) message_length: u16,
+}
+```
+
+### Methods
+
+```rust
+/// Since the following methods have very common names, `from_bytes` and `into_bytes`, the sole purpose of this struct
+/// is to give them a proper namespace to avoid confusion.
```

Comment:
Usually namespacing is done through module definitions. Can we not have a `tlv` module and put those static methods in there?

Reply:
Modules go away when you use the symbol. So there is no strict namespace.

Reply:
I don't get it, sorry. Why is
```rust
pub(crate) mod tlv {
   pub(crate) fn from_bytes<M: Message>(...) {}
   pub(crate) fn into_bytes<M: Message>(...) {} 
}
```
not possible?

Reply:
Because `use tlv:: from_bytes` doesn't keep `tlv`. Then you have a pretty common name in the scope which is weird. With the struct you just can't use it without `tlv::from_bytes` which is clearer.

Reply:
Why would you import `from_bytes` from the `tlv` module? Just use `tlv::from_bytes` wherever you need it?

Reply:
Why would you import anything from a module ? This is just avoiding a potential weird situation at no cost.

Reply:
I am trying to understand this: You want to be able to write:
`use tlv::Tlv` and then write:
`Tlv::from_bytes(...)` and `Tlv::into_bytes(...)`?

Reply:
`Tlv::from_bytes(...)` and `Tlv::into_bytes(...)` but  yes

Reply:
Ah you edited. No you don't need a struct at all.

Reply:
haha, I still don't get the issue. :joy:

Reply:
what's so favorable about `Tlv::from_bytes(...)` over `tlv::from_bytes(...)`?

Reply:
You can't get rid of the namespace.

Reply:
but you neither

Reply:
Of course you can, just importing it remove the namespace.

Reply:
Removed.